crash interaction between `extension def`, default args, and path-dependent typing

Compiler crash. This requires interaction between `extension def`, default arg being kicked in, and path-dependent typing.

## Compiler version

`3.8.*`
`3.7.*`
`3.6.4`
probably other versions too

setting `-source <version>` won't change that

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

[try this on Scastie..](https://scastie.scala-lang.org/WXk4dKYeQJiRAgfTCoOeSw)

```Scala

@deprecated("pls work this out")
def ??? = scala.Predef.`???`

import util.ChainingOps.{*, given}

object S {

trait AllocatedNodeTypeFamily:
  type Node

trait Parser[+A] private[S] () :
  type SpcNd

type OutliningParser[+A, +Node]
= Parser[A] { type SpcNd <: Node }

extension [R1] (g1: Parser[R1] )
              (using pfmoIspcndi: ParserFlatMapOp.ISpcNdInvariant = new ParserFlatMapOp.ISpcNdInvariant {}.asInstanceOf[ParserFlatMapOp.ISpcNdInvariant { type SpcNdObj = g1.SpcNd } ] )

            def     map[R2] (m: R1 => (       R2  ) ): OutliningParser[R2, g1.SpcNd] = new Parser[Any] {}.asInstanceOf
            def flatMap[R2] (m: R1 => (Parser[R2] ) ): OutliningParser[R2, g1.SpcNd] = new Parser[Any] {}.asInstanceOf

object ParserFlatMapOp {

  given given_ISpcNdInvariant_PModAllocatedNodeTypeFamily
  : (x: AllocatedNodeTypeFamily ) => (ISpcNdInvariant { type SpcNdObj = x.Node })
  = new ISpcNdInvariant {}.asInstanceOf

  trait ISpcNdInvariant :
    type SpcNdObj

} // ParserFlatMapOp.

def defaultIntParser
  : Parser[Int]
= new Parser[Any] {}.asInstanceOf

} // S.

System.err.println(Math.random() )

System.err.println(Math.random() )

locally :
  System.err.println(Math.random() )

  // given S.ParserFlatMapOp.ISpcNdInvariant
  // = (new S.ParserFlatMapOp.ISpcNdInvariant {}).asInstanceOf

  /* run locally you might see something like say "unexpected tree for type application" or the like.. 😒 */
  /* commented out, the whole code would compile and run like it should.. */
  locally :
    for
      r1 <- S.defaultIntParser
      r2 <- S.defaultIntParser
    yield { 5 }
  .toString()

System.err.println(s"done.")

System.err.println(Math.random() )

```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
java.lang.AssertionError: assertion failed: unexpected tree for type application: Block(List(ValDef(g1$1,TypeTree[AppliedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Playground$)),module class S$)),trait Parser),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int)))],Select(Ident(S),defaultIntParser))),Apply(Apply(TypeApply(Select(Ident(S),flatMap),List(TypeTree[TypeVar(TypeParamRef(R1))])),List(Ident(g1$1))),List(Apply(TypeApply(Select(Ident(S),flatMap$default$2),List(TypeTree[TypeVar(TypeParamRef(R1))])),List(Ident(g1$1))))))
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
	at dotty.tools.dotc.ast.tpd$.TypeApply(tpd.scala:58)
	at dotty.tools.dotc.ast.tpd$TreeOps$.appliedToTypeTrees$extension(tpd.scala:1034)
	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:5187)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:4399)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:4017)
	at dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:4668)
	at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:4693)
	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:4906)
	at dotty.tools.dotc.typer.Typer.adaptToArgs$1(Typer.scala:4492)
	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:5194)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:4399)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:4017)
	at dotty.tools.dotc.typer.Applications.extMethodApply(Applications.scala:2987)
	at dotty.tools.dotc.typer.Applications.extMethodApply$(Applications.scala:521)
	at dotty.tools.dotc.typer.Typer.extMethodApply(Typer.scala:163)
	at dotty.tools.dotc.typer.Implicits.tryExtension$1(Implicits.scala:1222)
	at dotty.tools.dotc.typer.Implicits.typedImplicit(Implicits.scala:1244)
	at dotty.tools.dotc.typer.Implicits.typedImplicit$(Implicits.scala:880)
	at dotty.tools.dotc.typer.Typer.typedImplicit(Typer.scala:163)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.tryImplicit(Implicits.scala:1323)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.rank$1(Implicits.scala:1493)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1663)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1787)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit$$anonfun$3(Implicits.scala:1802)
	at dotty.tools.dotc.typer.Implicits$SearchResult.recoverWith(Implicits.scala:434)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1788)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:1829)
	at dotty.tools.dotc.typer.Implicits.inferImplicit(Implicits.scala:1141)
	at dotty.tools.dotc.typer.Implicits.inferImplicit$(Implicits.scala:880)
	at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:163)
	at dotty.tools.dotc.typer.Implicits.inferView(Implicits.scala:920)
	at dotty.tools.dotc.typer.Implicits.inferView$(Implicits.scala:880)
	at dotty.tools.dotc.typer.Typer.inferView(Typer.scala:163)
	at dotty.tools.dotc.typer.Typer.tryExtensionOrConversion(Typer.scala:4339)
...
```

not everything could be brought here as the hover ran off the screen
</details>

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -1028,7 +1028,9 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
 
     /** The current tree applied to given type argument list: `tree[targs(0), ..., targs(targs.length - 1)]` */
     def appliedToTypeTrees(targs: List[Tree])(using Context): Tree =
-      if targs.isEmpty then tree else TypeApply(tree, targs)
+      if targs.isEmpty then tree else tree match
+        case Block(stmts, expr) if stmts.nonEmpty => Block(stmts, TypeApply(expr, targs))
+        case _ => TypeApply(tree, targs)
 
     /** Apply to `()` unless tree's widened type is parameterless */
     def ensureApplied(using Context): Tree =

--- a/tests/pos/26238-min.scala
+++ b/tests/pos/26238-min.scala
@@ -1,0 +1,20 @@
+trait Thing[T]
+trait G
+
+object Outer:
+
+  def thing: Thing[Int] = ???
+
+  extension[T1](t: Thing[T1])(using g: G = new G { })
+    def map[T2](m: T1 => T2): Thing[T2] = ???
+    def flatMap[T2](m: T1 => Thing[T2]): Thing[T2] = ???
+
+  // `y` requires a `flatMap`, and that one requires getting the default value for the `g` argument,
+  // and getting that default value requires the `t` argument (since default values can depend on previous args),
+  // which means we end up with a block;
+  // and since the whole thing is generic, we must apply a generic type argument to that block
+  for
+    x <- thing
+    y <- thing
+  yield
+    1

--- a/tests/pos/26238.scala
+++ b/tests/pos/26238.scala
@@ -1,0 +1,65 @@
+object toplevel {
+
+@deprecated("pls work this out")
+def ??? = scala.Predef.`???`
+
+import util.ChainingOps.{*, given}
+
+object S {
+
+trait AllocatedNodeTypeFamily:
+  type Node
+
+trait Parser[+A] private[S] () :
+  type SpcNd
+
+type OutliningParser[+A, +Node]
+= Parser[A] { type SpcNd <: Node }
+
+extension [R1] (g1: Parser[R1] )
+              (using pfmoIspcndi: ParserFlatMapOp.ISpcNdInvariant = new ParserFlatMapOp.ISpcNdInvariant {}.asInstanceOf[ParserFlatMapOp.ISpcNdInvariant { type SpcNdObj = g1.SpcNd } ] )
+
+            def     map[R2] (m: R1 => (       R2  ) ): OutliningParser[R2, g1.SpcNd] = new Parser[Any] {}.asInstanceOf
+            def flatMap[R2] (m: R1 => (Parser[R2] ) ): OutliningParser[R2, g1.SpcNd] = new Parser[Any] {}.asInstanceOf
+
+object ParserFlatMapOp {
+
+  given given_ISpcNdInvariant_PModAllocatedNodeTypeFamily
+  : (x: AllocatedNodeTypeFamily ) => (ISpcNdInvariant { type SpcNdObj = x.Node })
+  = new ISpcNdInvariant {}.asInstanceOf
+
+  trait ISpcNdInvariant :
+    type SpcNdObj
+
+} // ParserFlatMapOp.
+
+def defaultIntParser
+  : Parser[Int]
+= new Parser[Any] {}.asInstanceOf
+
+} // S.
+
+System.err.println(Math.random() )
+
+System.err.println(Math.random() )
+
+locally :
+  System.err.println(Math.random() )
+
+  // given S.ParserFlatMapOp.ISpcNdInvariant
+  // = (new S.ParserFlatMapOp.ISpcNdInvariant {}).asInstanceOf
+
+  /* run locally you might see something like say "unexpected tree for type application" or the like.. 😒 */
+  /* commented out, the whole code would compile and run like it should.. */
+  locally :
+    for
+      r1 <- S.defaultIntParser
+      r2 <- S.defaultIntParser
+    yield { 5 }
+  .toString()
+
+System.err.println(s"done.")
+
+System.err.println(Math.random() )
+
+}

--- a/tests/pos/26238.scala
+++ b/tests/pos/26238.scala
@@ -1,8 +1,5 @@
 object toplevel {
 
-@deprecated("pls work this out")
-def ??? = scala.Predef.`???`
-
 import util.ChainingOps.{*, given}
 
 object S {
@@ -39,27 +36,12 @@ def defaultIntParser
 
 } // S.
 
-System.err.println(Math.random() )
-
-System.err.println(Math.random() )
-
 locally :
-  System.err.println(Math.random() )
-
-  // given S.ParserFlatMapOp.ISpcNdInvariant
-  // = (new S.ParserFlatMapOp.ISpcNdInvariant {}).asInstanceOf
-
-  /* run locally you might see something like say "unexpected tree for type application" or the like.. 😒 */
-  /* commented out, the whole code would compile and run like it should.. */
   locally :
     for
       r1 <- S.defaultIntParser
       r2 <- S.defaultIntParser
     yield { 5 }
   .toString()
-
-System.err.println(s"done.")
-
-System.err.println(Math.random() )
 
 }

